lib/searches.js

```diff
@@ -129,8 +130,95 @@ var breadthFS = function (url, searchDepth) {
     return searchingPromise;
 }
 
-var depthFS = function () {
-    // To be implemented
+var dfsHelper = function (url, edges, level) {
+    // Get URL and create promise
+    var helperPromise = request({url: url, jar: true});
+    helperPromise.then((body) => {
+        var r = new RegExp('^https?://', 'i');
+        var numInvalidLinks = 0;
+        var randomUrl = "";
+        var isValid = false;
+        
+        // Load HTML and get all links in page
+        var $ = cheerio.load(body);;
+        var sourceTitle = $("title").text();
+        console.log("Link <title>", sourceTitle);
+        links = $('a');
+      
+        // Loop through random page links until an absolute link that looks valid is found.
+        while (!isValid){  
+            if (((typeof(randomUrl) == 'undefined' || !validUrl.isUri(randomUrl)) || !r.test(randomUrl)) && numInvalidLinks < 50) {
+                randomUrl = $($(links).get(Math.floor(Math.random() * $(links).length))).attr("href");
+                numInvalidLinks++;
+                continue;   
+            }
+            // If page loops for valid links 50 times, conclude there are none and return.
+            if(numInvalidLinks == 50){
+                console.log("No more links. Sending results...");
+                break;
+            }     
+            // If link looks valid, add to edge list and return.   
+            isValid = true;
+            numInvalidLinks = 0;
+            edges[level] = ({
+                "SourceUrl": url,
+                "SourceTitle": sourceTitle,
+                "DestinationUrl": randomUrl,
+                "Level": level + 1
+            });
+        }                 
+    });
+    
+    helperPromise.catch((err) => {
+        throw err;
+    });
+    
+    return helperPromise;
+}
+
+// Note on searchDepth parameter:
+// * A searchDepth of 0 terminates immediately.
+// * A searchDepth of n performs a search depth of n, beginning at the rootUrl
+//   and following random nodes until a depth of n is reached.
+var depthFS = function (rootUrl, searchDepth) {
+    var currentDepth = 0;
+    var edges = [];
+    var nextUrl = rootUrl;
+
+    var searchingPromise = new Promise((resolve, reject) => {
+        console.log("Crawling level 0...");
+
+        // Get random links until the searchDepth is reached.
+        function getRandomLink(url) {
+            if (currentDepth >= searchDepth) {
+                resolve(edges);
+                return;
+            }
+           
+            var helperPromise = dfsHelper(nextUrl, edges, currentDepth);
+            
+            helperPromise.then(() => {
+                // If link error, return results thus far.
+                if (edges[currentDepth] === undefined){
+                    resolve(edges);
```

Comment:
I see what you mean about hitting a dead end and returning. I think it's fine to return a shorter list of edges than requested. Maybe for extra credit we could mark the last node as the "dead end" somehow. 

If we really wanted to search further, then we can return a promise inside a `then` or `catch` handler to prevent the handler from resolving until the inner promise resolves. (Technically, the inner promise gets merged into the outer promise, with its internal queue placed before the outer promise's queue. That way the inner promise has so complete all of it's handlers before the outer queue can go on). We would decrement `currentDepth` and call `getRandomLink` on the previous link again, trying to get a new link which has more results. I don't think this is necessary now, but maybe we add it as another stretch goal.

Reply:
I was thinking that I would add a check here using a head request and if it were a dead end or 404, I would have it try again on a different random link. I was having so much trouble with the outer promise not waiting on the inner promise though. So if I return a promise inside a then or catch it should work?  Do you think that a head request to check for 404 is the right way to go about it?